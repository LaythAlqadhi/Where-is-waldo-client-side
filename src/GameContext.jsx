import React, { useState, createContext } from 'react';
import PropTypes from 'prop-types';

const initialState = {
  state: { loading: false, error: null },
  status: { isStarted: false, isEnded: true },
  message: null,
};

export const GameContext = createContext(initialState);

export function GameProvider({ children }) {
  const [state, setState] = useState({ loading: false, error: null });
  const [status, setStatus] = useState({
    isStarted: false,
    isEnded: false,
    message: null,
  });
  const [message, setMessage] = useState(null);

  return (
    <GameContext.Provider
      value={{
        state,
        setState,
        status,
        setStatus,
        message,
        setMessage,
      }}
    >
      {children}
    </GameContext.Provider>
  );
}

GameProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
