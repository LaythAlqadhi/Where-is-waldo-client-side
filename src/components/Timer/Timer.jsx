import React, { useState, useEffect, useContext } from 'react';
import { GameContext } from '../../GameContext';

function Timer() {
  const { status } = useContext(GameContext);
  const [time, setTime] = useState(0);

  useEffect(() => {
    let interval;

    if (status.isStarted) {
      interval = setInterval(() => {
        setTime((prevTime) => prevTime + 1);
      }, 1000);
    } else {
      setTime(0);
    }

    return () => clearInterval(interval);
  }, [status.isStarted]);

  const formatTime = (seconds) => {
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    const remainingSeconds = seconds % 60;

    const formattedTime = `${String(hours).padStart(2, '0')}:${String(
      minutes,
    ).padStart(2, '0')}:${String(remainingSeconds).padStart(2, '0')}`;

    return formattedTime;
  };

  return <span aria-label="Timer">{formatTime(time)}</span>;
}

export default Timer;
