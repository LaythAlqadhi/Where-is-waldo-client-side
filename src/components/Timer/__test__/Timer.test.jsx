import React from 'react';
import { render } from '@testing-library/react';
import { GameContext } from '../../../GameContext';
import Timer from '../Timer';

it('renders Timer component correctly', () => {
  const { asFragment } = render(
    <GameContext.Provider
      value={{ status: { isStarted: true, isEnded: false } }}
    >
      <Timer />
    </GameContext.Provider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
