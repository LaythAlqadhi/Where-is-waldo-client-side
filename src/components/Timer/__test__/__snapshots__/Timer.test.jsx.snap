// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Timer component correctly 1`] = `
<DocumentFragment>
  <span
    aria-label="Timer"
  >
    00:00:00
  </span>
</DocumentFragment>
`;
